lib/Check.vala

```diff
@@ -36,13 +35,52 @@ public abstract class ValaLint.Check {
     public abstract string get_description ();
 
     /**
-     * Checks a given line for formatting mistakes.
+     * Checks a given parse result for formatting mistakes.
      *
+     * @param parse_result The parsed string.
      * @param mistake_list The list new mistakes should be added to.
-     * @param line_index The index of the given line.
-     * @param line The line that should be checked.
+     */
+    public abstract void check (Gee.ArrayList<ParseResult?> parse_result, Gee.ArrayList<FormatMistake?> mistake_list);
+
+    /**
+     * Adds a mistake based on a regex pattern.
      *
-     * @return Indicates if new mistakes were added to the list.
+     * @param check The used check.
+     * @param pattern The regex pattern.
+     * @param mistake The mistake description.
+     * @param parse_result The current parse result element.
+     * @param mistakes The mistakes list.
+     * @param char_offset The offset between the mistake char position and the regex pattern start.
      */
-    public abstract bool check_line (Gee.ArrayList<FormatMistake?> mistake_list, int line_index, string line);
+    public static void add_regex_mistake (Check check, string pattern, string mistake, ParseResult parse_result, Gee.ArrayList<FormatMistake? > mistakes, int char_offset = 0, bool single_mistake_in_line = false) {
+
+        MatchInfo match_info;
+        try {
+            var regex = new Regex (pattern);
+            regex.match (parse_result.text, 0, out match_info);
+            while (match_info.matches ()) {
+                int pos_start, pos_end;
+                match_info.fetch_pos (0, out pos_start, out pos_end);
+
+                int pos_mistake = pos_start + char_offset;
+                int line_count = Utils.get_line_count (parse_result.text[0:pos_mistake]);
+                int line_pos = parse_result.line_pos + line_count;
+                int char_pos = Utils.get_char_index_in_line (parse_result.text, pos_mistake);
+                if (line_count == 0) {
+                    char_pos += parse_result.char_pos;
+                }
+
+                /* If single_mistake_in_line is true, add only one mistake of the same check per line */
+                if (!single_mistake_in_line ||
+                    (mistakes.is_empty || mistakes.last ().check != check || mistakes.last ().line_index < line_pos)) {
+                    mistakes.add ({ check, line_pos, char_pos, mistake});
+                }
+
+
+                match_info.next ();
+            }
+        } catch {
+            error (_("%s is not a valid Regex"), pattern);
```

Comment:
Errors do not need to be translatable, so you can remove the `_()` here again. Only strings that the user is going to see and use need to be translatable. Error messages are usually sent to the developer for debugging, so can remain english.

Also I think you want to use `critical` instead of `error` here. `error` closes the application, I think it would be best to continue the other checks even if this regex fails.